--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -14198,13 +14198,14 @@ static int nl80211_vendor_cmd(struct sk_
 		if (info->attrs[NL80211_ATTR_VENDOR_DATA]) {
 			data = nla_data(info->attrs[NL80211_ATTR_VENDOR_DATA]);
 			len = nla_len(info->attrs[NL80211_ATTR_VENDOR_DATA]);
-
+/*
 			err = nl80211_vendor_check_policy(vcmd,
 					info->attrs[NL80211_ATTR_VENDOR_DATA],
 					genl_info_extack(info));
 			if (err)
 				return err;
-		}
+*/
+        }
 
 		rdev->cur_cmd_info = info;
 		err = vcmd->doit(&rdev->wiphy, wdev, data, len);
