--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -3165,7 +3165,8 @@ static int nl80211_set_wiphy(struct sk_b
 		}
 
 		if (netdev->ieee80211_ptr->iftype != NL80211_IFTYPE_AP &&
-		    netdev->ieee80211_ptr->iftype != NL80211_IFTYPE_P2P_GO) {
+		    netdev->ieee80211_ptr->iftype != NL80211_IFTYPE_P2P_GO &&
+			netdev->ieee80211_ptr->iftype != NL80211_IFTYPE_STATION) {
 			result = -EINVAL;
 			goto out;
 		}