--- a/root/sbin/fw4
+++ b/root/sbin/fw4
@@ -33,6 +33,12 @@
 		ACTION=start \
 			ucode -S -i $MAIN | nft $VERBOSE -f /proc/self/fd/0
 	} 1000>$LOCK
+
+        OFFLOAD=$(uci -q get firewall.@defaults[0].flow_offloading_hw || exit 0)
+        [ $OFFLOAD -ne 0 ] || exit 0
+        WLAN_DEVS=$(iw dev | awk '/Interface/ {print $2}' | tr '\n' ',' | sed 's/,$//')
+        [ -n "$WLAN_DEVS" ] || exit 0
+        nft add flowtable inet fw4 ft { hook ingress priority 0 \; devices = { "$WLAN_DEVS" } \; flags offload \; } && logger -t firewall "Added $DEVICENAME to flowtable due to reload"
 }

 print() {
