--- a/src/ap/ap_config.c
+++ b/src/ap/ap_config.c
@@ -277,7 +277,7 @@ struct hostapd_config * hostapd_config_d
 		HE_OPERATION_RTS_THRESHOLD_OFFSET;
 	/* Set default basic MCS/NSS set to single stream MCS 0-7 */
 	conf->he_op.he_basic_mcs_nss_set = 0xfffc;
-	conf->he_op.he_bss_color_disabled = 1;
+	conf->he_op.he_bss_color_disabled = 0;
 	conf->he_op.he_bss_color_partial = 0;
 	conf->he_op.he_bss_color = os_random() % 63 + 1;
 	conf->he_op.he_twt_responder = 1;
