--- a/src/ap/ap_config.c
+++ b/src/ap/ap_config.c
@@ -189,7 +189,7 @@ struct hostapd_config * hostapd_config_d
 	const struct hostapd_wmm_ac_params ac_vo = /* voice traffic */
 		{ aCWmin - 2, aCWmin - 1, 2, 1504 / 32, 0 };
 	const struct hostapd_tx_queue_params txq_bk =
-		{ 7, ecw2cw(aCWmin), ecw2cw(aCWmax), 0 };
+		{ 3, ecw2cw(aCWmin), ecw2cw(aCWmax), 0 };
 	const struct hostapd_tx_queue_params txq_be =
 		{ 3, ecw2cw(aCWmin), 4 * (ecw2cw(aCWmin) + 1) - 1, 0};
 	const struct hostapd_tx_queue_params txq_vi =
