--- a/src/ap/drv_callbacks.c
+++ b/src/ap/drv_callbacks.c
@@ -945,23 +945,29 @@ void hostapd_event_ch_switch(struct host
 	hapd->iconf->ieee80211n = ht;
 	if (!ht) {
 		hapd->iconf->ieee80211ac = 0;
-	} else if (hapd->iconf->ch_switch_vht_config) {
-		/* CHAN_SWITCH VHT config */
-		if (hapd->iconf->ch_switch_vht_config &
-		    CH_SWITCH_VHT_ENABLED)
-			hapd->iconf->ieee80211ac = 1;
-		else if (hapd->iconf->ch_switch_vht_config &
-			 CH_SWITCH_VHT_DISABLED)
-			hapd->iconf->ieee80211ac = 0;
-	} else if (hapd->iconf->ch_switch_he_config) {
-		/* CHAN_SWITCH HE config */
-		if (hapd->iconf->ch_switch_he_config &
-		    CH_SWITCH_HE_ENABLED)
-			hapd->iconf->ieee80211ax = 1;
-		else if (hapd->iconf->ch_switch_he_config &
-			 CH_SWITCH_HE_DISABLED)
-			hapd->iconf->ieee80211ax = 0;
-	}
+		hapd->iconf->ieee80211ax = 0;
+	} else {
+		if (hapd->iconf->ch_switch_vht_config) {
+			/* CHAN_SWITCH VHT config */
+			if (hapd->iconf->ch_switch_vht_config &
+			    CH_SWITCH_VHT_ENABLED)
+				hapd->iconf->ieee80211ac = 1;
+			else if (hapd->iconf->ch_switch_vht_config &
+				 CH_SWITCH_VHT_DISABLED)
+				hapd->iconf->ieee80211ac = 0;
+		}
+		if (hapd->iconf->ch_switch_he_config) {
+			/* CHAN_SWITCH HE config */
+			if (hapd->iconf->ch_switch_he_config &
+			    CH_SWITCH_HE_ENABLED) {
+				hapd->iconf->ieee80211ac = 1;
+				hapd->iconf->ieee80211ax = 1;
+			} else if (hapd->iconf->ch_switch_he_config &
+				 CH_SWITCH_HE_DISABLED) {
+				hapd->iconf->ieee80211ax = 0;
+			}
+		}
+ 	}
 	hapd->iconf->ch_switch_vht_config = 0;
 	hapd->iconf->ch_switch_he_config = 0;
 
