Index: hostapd-2022-01-16-cff80b4f/src/ap/beacon.c
===================================================================
--- hostapd-2022-01-16-cff80b4f.orig/src/ap/beacon.c
+++ hostapd-2022-01-16-cff80b4f/src/ap/beacon.c
@@ -556,7 +556,8 @@ static u8 * hostapd_gen_probe_resp(struc
 #endif /* CONFIG_FST */

 #ifdef CONFIG_IEEE80211AC
-	if (hapd->iconf->ieee80211ac && !hapd->conf->disable_11ac &&
+	if ((hapd->iconf->channel > 14) &&
+        hapd->iconf->ieee80211ac && !hapd->conf->disable_11ac &&
 	    !is_6ghz_op_class(hapd->iconf->op_class)) {
 		pos = hostapd_eid_vht_capabilities(hapd, pos, 0);
 		pos = hostapd_eid_vht_operation(hapd, pos);
Index: hostapd-2022-01-16-cff80b4f/src/ap/ieee802_11.c
===================================================================
--- hostapd-2022-01-16-cff80b4f.orig/src/ap/ieee802_11.c
+++ hostapd-2022-01-16-cff80b4f/src/ap/ieee802_11.c
@@ -5136,7 +5136,8 @@ static u16 send_assoc_resp(struct hostap
 	p = hostapd_eid_ht_operation(hapd, p);

 #ifdef CONFIG_IEEE80211AC
-	if (hapd->iconf->ieee80211ac && !hapd->conf->disable_11ac &&
+	if ((hapd->iconf->channel > 14) &&
+        hapd->iconf->ieee80211ac && !hapd->conf->disable_11ac &&
 	    !is_6ghz_op_class(hapd->iconf->op_class)) {
 		u32 nsts = 0, sta_nsts;

