--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -1512,9 +1512,8 @@ static int hostapd_setup_bss(struct host
 }
 
 
-static void hostapd_tx_queue_params(struct hostapd_iface *iface)
+static void hostapd_tx_queue_params(struct hostapd_iface *iface, struct hostapd_data *hapd)
 {
-	struct hostapd_data *hapd = iface->bss[0];
 	int i;
 	struct hostapd_tx_queue_params *p;
 
@@ -2235,11 +2234,10 @@ static int hostapd_setup_interface_compl
 		}
 		if (is_zero_ether_addr(hapd->conf->bssid))
 			prev_addr = hapd->own_addr;
+		hostapd_tx_queue_params(iface, hapd);
 	}
 	hapd = iface->bss[0];
 
-	hostapd_tx_queue_params(iface);
-
 	ap_list_init(iface);
 
 	hostapd_set_acl(hapd);

--- a/wpa_supplicant/driver_i.h
+++ b/wpa_supplicant/driver_i.h
@@ -10,6 +10,7 @@
 #define DRIVER_I_H
 
 #include "drivers/driver.h"
+#include "config.h"
 
 /* driver_ops */
 static inline void * wpa_drv_init(struct wpa_supplicant *wpa_s,
@@ -287,6 +288,28 @@ static inline int wpa_drv_set_operstate(
 	return 0;
 }
 
+static inline int wpa_drv_set_tx_queue_params(struct wpa_supplicant *wpa_s,
+	                                              int queue, int aifs, int cw_min,
+	                                              int cw_max, int burst_time)
+{
+	if (wpa_s->driver->set_tx_queue_params)
+		return wpa_s->driver->set_tx_queue_params(wpa_s->drv_priv, queue, aifs, cw_min, cw_max, burst_time);
+	return 0;
+}
+	
+static void wpa_drv_tx_queue_params(struct wpa_supplicant *wpa_s)
+{
+	struct hostapd_tx_queue_params *p;
+	int i;
+
+	for (i = 0; i < NUM_TX_QUEUES; i++) {
+		p = &wpa_s->conf->tx_queue[i];
+
+		if (wpa_drv_set_tx_queue_params(wpa_s, i, p->aifs, p->cwmin, p->cwmax, p->burst))
+			wpa_printf(MSG_DEBUG, "Failed to set TX queue parameters for queue %d.", i);
+	    }
+}
+
 static inline int wpa_drv_mlme_setprotection(struct wpa_supplicant *wpa_s,
 					     const u8 *addr, int protect_type,
 					     int key_type)

--- a/wpa_supplicant/wpa_supplicant.c
+++ b/wpa_supplicant/wpa_supplicant.c
@@ -6736,6 +6736,8 @@ static int wpa_supplicant_init_iface(str
 	if (wpa_supplicant_init_wpa(wpa_s) < 0)
 		return -1;
 
+	wpa_drv_tx_queue_params(wpa_s);
+
 	wpa_sm_set_ifname(wpa_s->wpa, wpa_s->ifname,
 			  wpa_s->bridge_ifname[0] ? wpa_s->bridge_ifname :
 			  NULL);
